### PR TITLE
[WIP] Headless training

### PR DESCRIPTION
**Pull request by jjerphan**
Main changes:
 - update script w.r.t latest changes in the codebase
 - adapt it for Autocontext
 - adapt CLI

### CODE REVIEW

**Comment by jjerphan** [comment]
Adapting this script to support PixelClassification and Autocontext workflows (including the 3- and 4-stages ones) is relatively easy. It is rather a problem of having an intuitive interface.

In term of the CPI, my usage would be:

```bash
train_headless.py project_to_create.ilp \
  --gaussianSmoothing=0.05,0.2,0.3 \
  --laplacian_of_gaussian=0.1 \
  # ditto for other FeatureIds 
  --workflow="Autocontext (2-stage)" \
  "/path/to/training_data/*.h5" \
  "/path/to/labels/*.h5"
```

I would appreciate suggestions. :slightly_smiling_face:

**Comment by jjerphan** [comment]
I think that #2303 introduces an API which addresses similar needs for a programmatic way of interaction with Ilastik.

@k-dominik, do you have any remarks regarding such a contribution? :slightly_smiling_face:

**Comment by k-dominik** [comment]
Hi @jjerphan,

you are indeed right, that we are introducing a python api with #2303. Going relatively slow in baby-steps there. The next step would be enabling prediction for autocontext, too and then going to the training part, where the most important decision would be how to handle inputs.

In the meantime this script should be fixed, so your changes are very welcome :). However, further development with this should follow, once the relevant parts of the api are released.

Do you want to make this PR ready for review?

Cheers
Dominik

**Comment by jjerphan** [comment]
Hi @k-dominik,

Thanks for the feedback. I prefer waiting once you came with a good python interface, to then find a proper CLI for this PR.

Julien.

**Comment by jjerphan** [comment]
Hi @k-dominik
Are you willing to reconsider this PR as #2303 has been merged? :slightly_smiling_face:

File: ilastik/applets/dataSelection/dataSelectionApplet.py [hunk 1]
```diff
@@ -241,7 +241,7 @@ def configure_operator_with_parsed_args(self, parsed_args):
 
                 input_paths = self.convertStacksToH5(input_paths, tempfile.gettempdir())
 
-            input_infos = [FilesystemDatasetInfo(filepath=p) if p else None for p in input_paths]
+            input_infos = [FilesystemDatasetInfo(filePath=p) if p else None for p in input_paths]
```
Comment (k-dominik) [inline ilastik/applets/dataSelection/dataSelectionApplet.py:244]: Hey, this part has been reworked completely in the meantime :/

File: bin/train_headless.py [hunk 2]
```diff
@@ -154,7 +154,8 @@ def generate_trained_project_file(
     data_selection_applet = shell.workflow.dataSelectionApplet
 
     # To configure data selection, start with empty cmdline args and manually fill them in
-    data_selection_args, _ = data_selection_applet.parse_known_cmdline_args([], PixelClassificationWorkflow.ROLE_NAMES)
+    data_selection_args, _ = data_selection_applet.parse_known_cmdline_args(cmdline_args=[],
+                                                                            role_names=["Raw Data", "Prediction Mask"])
```
Comment (k-dominik) [inline bin/train_headless.py:158]: Hey I think this has also changed slightly in the meantime.. Here I'd suggest something like

```Python
...
role_names=[x.name for x in PixelClassificationWorkflow.Roles]
```

**Review by k-dominik — COMMENTED** [review]
Hey @jjerphan,

thank you very much for fixing this script and I'm extremely sorry that we've been so slow to react here. In the meantime parts of the code you are accessing have changed :/

I'd suggest to start fixing this from the current master (sorry)...

**Comment by jjerphan** [comment]
> Hey @jjerphan,
> 
> thank you very much for fixing this script and I'm extremely sorry that we've been so slow to react here. In the meantime parts of the code you are accessing have changed :/

Hi @k-dominik 
It is fine. This change is relatively minimal and was a quick fix for me.
I am not actively working on the project I used Ilastik for, but I may in the future.
 
> we've also not yet agreed on a Python API for training :/

I guess a CLI would make more sense for training as well. In this context, I can definitely wait for a dedicated python API once it's set. I would be happy to review this Python API, let me know if you are interested. 🙂

**Pull request closed without merging**